janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/HasStepFolder.java

```diff
@@ -227,6 +226,33 @@ public OrderEntry(String key, Order order) {
             this.key = key;
             this.order = order;
         }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            OrderEntry that = (OrderEntry) o;
+
+            if (key != null ? !key.equals(that.key) : that.key != null) return false;
+            return order == that.order;
+
+        }
+
+        @Override
+        public int hashCode() {
+            int result = key != null ? key.hashCode() : 0;
+            result = 31 * result + (order != null ? order.hashCode() : 0);
```

Comment:
Might be more readable as:

```java
int keyCode = (key != null) ? key.hashCode() : 0;
int orderCode = (order != null) ? order.hashCode() : 0;
return (31 * keyCode) + orderCode;
```

Because it crease parallel structure for the two elements.